File: UseCasesCore/Sources/UseCasesCore/Markers/Commands/BindingLifecycle.swift
/// How a marker is placed.
public enum MarkerMode: String, Equatable, Sendable {
  case explicit
  case swiftFunction = "swift-func"
}

/// Where a marker should go: a single declaration line (swift-func, the marker
/// goes immediately before it) or an inclusive span (explicit, a marker on each
/// side).
public struct MarkerPlacement: Equatable, Sendable {
  public let mode: MarkerMode
  public let line: Int?
  public let startLine: Int?
  public let endLine: Int?

  public init(
    mode: MarkerMode,
    line: Int? = nil,
    startLine: Int? = nil,
    endLine: Int? = nil,
  ) {
    self.mode = mode
    self.line = line
    self.startLine = startLine
    self.endLine = endLine
  }
}

/// One marker occurrence in one file. `endLine` is nil for a lone start marker.
public struct MarkerLocation: Equatable, Sendable {
  public let filePath: String
  public let startLine: Int
  public let endLine: Int?

  public init(
    filePath: String,
    startLine: Int,
    endLine: Int?,
  ) {
    self.filePath = filePath
    self.startLine = startLine
    self.endLine = endLine
  }

  /// `{ file_path, start_line, end_line }`, `end_line` null when absent.
  var jsonValue: JSONValue {
    .object(JSONObject([
      ("file_path", .string(filePath)),
      ("start_line", .number(Double(startLine))),
      ("end_line", JSONValue.optionalNumber(endLine)),
    ]))
  }
}

/// The edited contents, or why the edit was refused.
public enum MarkerInsertResult: Equatable, Sendable {
  case contents(String)
  case refused(MarkerCommandFailure)

  /// `{ contents }` or `{ error: { code, message } }`.
  var jsonValue: JSONValue {
    switch self {
    case let .contents(contents):
      .object(JSONObject([("contents", .string(contents))]))
    case let .refused(failure):
      .object(JSONObject([("error", failure.jsonValue)]))
    }
  }
}

/// A slug's marker as found in the product tree, with its file's contents.
public struct FoundSlugMarker: Equatable, Sendable {
  public let location: MarkerLocation
  public let contents: String
  public let commentPrefix: String
}

/// The inputs of a registry event.
public struct RegistryEventInput: Equatable, Sendable {
  public let command: String
  public let rowIdentifier: String
  public let bindingSlug: String
  public let reason: String
  public let eventIdentifier: String
  public let createdAt: String
  /// Nil records ``ProductVersion/version``.
  public let version: String?

  public init(
    command: String,
    rowIdentifier: String,
    bindingSlug: String,
    reason: String,
    eventIdentifier: String,
    createdAt: String,
    version: String? = nil,
  ) {
    self.command = command
    self.rowIdentifier = rowIdentifier
    self.bindingSlug = bindingSlug
    self.reason = reason
    self.eventIdentifier = eventIdentifier
    self.createdAt = createdAt
    self.version = version
  }
}

/// The marker editing and registry-event plumbing shared by the three commands
/// that write a binding (bindingLifecycle.ts): bind places and registers,
/// unbind removes and releases, rebind moves, releases and re-registers.
///
/// Kept in one place because the tool must be the single writer of BOTH halves
/// of a binding — the marker in the source and the event in the registry.
public enum BindingLifecycle {
  /// `insertMarkerLines`: the marker(s) for `slug` inserted at `placement`.
  ///
  /// Lines are split on LF ONLY, so a CRLF source keeps its carriage return on
  /// every line while the inserted marker line has none — mixed endings, as the
  /// TypeScript writes them. The marker is never indented. For an explicit span
  /// the end marker goes in first, at the higher index, so inserting the start
  /// does not shift it.
  public static func insertMarkerLines(
    source: String,
    commentPrefix: String,
    slug: String,
    placement: MarkerPlacement,
  ) -> MarkerInsertResult {
    var (lines, terminator) = splitKeepingTerminator(source)
    let marker = markerToken(commentPrefix) + slug

    switch placement.mode {
    case .swiftFunction:
      guard let line = placement.line, line >= 1 else {
        return .refused(MarkerCommandFailure(
          code: "BIND_LINE_REQUIRED",
          message: "--line is required for swift-func bind",
        ))
      }
      guard line - 1 <= lines.count else {
        return .refused(MarkerCommandFailure(
          code: "BIND_LINE_OUT_OF_RANGE",
          message: "--line \(line) is past end of file",
        ))
      }
      lines.insert(marker, at: line - 1)
      return .contents(joinWithTerminator(lines, terminator))
    case .explicit:
      guard let startLine = placement.startLine, let endLine = placement.endLine else {
        return .refused(MarkerCommandFailure(
          code: "BIND_SPAN_REQUIRED",
          message: "--start-line and --end-line are required for explicit bind",
        ))
      }
      guard startLine >= 1, endLine >= startLine, endLine <= lines.count else {
        return .refused(MarkerCommandFailure(
          code: "BIND_SPAN_OUT_OF_RANGE",
          message: "explicit span \(startLine)-\(endLine) is out of range",
        ))
      }
      lines.insert(markerToken(commentPrefix) + "end " + slug, at: endLine)
      lines.insert(marker, at: startLine - 1)
      return .contents(joinWithTerminator(lines, terminator))
    }
  }

  /// `locateMarkerLines`: the first start marker for `slug` and the first end
  /// marker for it after that start, read from the raw lines rather than the
  /// scanner's records — a marker whose span is BROKEN produces no record, and
  /// that is exactly the state someone needs to release.
  public static func locateMarkerLines(
    filePath: String,
    contents: String,
    commentPrefix: String,
    slug: String,
  ) -> MarkerLocation? {
    let (lines, _) = splitKeepingTerminator(contents)
    var startIndex: Int?
    var endIndex: Int?
    for (index, line) in lines.enumerated() {
      switch MarkerLineParser.parse(line, commentPrefix: commentPrefix) {
      case let .start(parsedSlug, _, _)
        where JavaScriptString.identical(parsedSlug, slug) && startIndex == nil:
        startIndex = index
      case let .end(parsedSlug, _)
        where JavaScriptString.identical(parsedSlug, slug) && startIndex != nil && endIndex == nil:
        endIndex = index
      default:
        continue
      }
    }
    guard let startIndex else {
      return nil
    }
    return MarkerLocation(
      filePath: filePath,
      startLine: startIndex + 1,
      endLine: endIndex.map { $0 + 1 },
    )
  }

  /// `removeMarkerLines`: the lines `location` names dropped; a line number
  /// outside the file drops nothing.
  public static func removeMarkerLines(
    contents: String,
    location: MarkerLocation,
  ) -> String {
    let (lines, terminator) = splitKeepingTerminator(contents)
    var dropped: Set<Int> = [location.startLine - 1]
    if let endLine = location.endLine {
      dropped.insert(endLine - 1)
    }
    let kept = lines.enumerated()
      .filter { line in
        !dropped.contains(line.offset)
      }
      .map(\.element)
    return joinWithTerminator(kept, terminator)
  }

  /// `findSlugMarker`: the slug's marker wherever it sits in the product tree,
  /// in the walk's sorted order, with its file's contents; nil when no file
  /// carries one.
  public static func findSlugMarker(
    files: some MarkerFileSystem,
    productRoot: String,
    slug: String,
    commentConfiguration: CommentPrefixConfiguration? = nil,
    skipPaths: [String] = [],
  ) throws(FileAccessError) -> FoundSlugMarker? {
    let inputs = try MarkerCommandInputs.collectSourceInputs(
      productRoot: productRoot,
      files: files,
      configuration: commentConfiguration,
      skipPaths: skipPaths,
    )
    for input in inputs {
      guard let commentPrefix = CommentPrefix.resolve(
        filePath: input.filePath,
        configuration: commentConfiguration,
        contents: input.contents,
      ) else {
        continue
      }
      if let location = locateMarkerLines(
        filePath: MarkerCommandInputs.toPosix(input.filePath),
        contents: input.contents,
        commentPrefix: commentPrefix,
        slug: slug,
      ) {
        return FoundSlugMarker(
          location: location,
          contents: input.contents,
          commentPrefix: commentPrefix,
        )
      }
    }
    return nil
  }

  /// `bindingRegisteredEvent`.
  public static func bindingRegisteredEvent(_ input: RegistryEventInput) -> JSONValue {
    registryEvent(.bindingRegistered, input)
  }

  /// `bindingReleasedEvent`: the event that ENDS a registration — appended,
  /// never a deletion, so the registry stays append-only.
  public static func bindingReleasedEvent(_ input: RegistryEventInput) -> JSONValue {
    registryEvent(.bindingReleased, input)
  }

  /// The event in the TypeScript's key order. `tool` is the literal
  /// `use-cases`: it is ledger data, frozen with the ledger format.
  private static func registryEvent(
    _ eventType: RegistryEventType,
    _ input: RegistryEventInput,
  ) -> JSONValue {
    .object(JSONObject([
      ("schema", .string(MarkerConstants.bindingRegistrySchemaIdentifier)),
      ("event_type", .string(eventType.rawValue)),
      ("event_id", .string(input.eventIdentifier)),
      ("created_at", .string(input.createdAt)),
      ("created_by", .object(JSONObject([
        ("tool", .string("use-cases")),
        ("command", .string(input.command)),
        ("version", .string(input.version ?? ProductVersion.version)),
      ]))),
      ("row_id", .string(input.rowIdentifier)),
      ("binding_slug", .string(input.bindingSlug)),
      ("reason", .string(input.reason)),
    ]))
  }

  /// `<prefix>: @use-case:` — spelled in two pieces so this source file never
  /// carries a marker line.
  private static func markerToken(_ commentPrefix: String) -> String {
    commentPrefix + ": @use-" + "case:"
  }

  /// Logical lines split on LF code units, and whether the source ended with
  /// one. Worked in code units: a Swift `Character` would fold `\r\n` into one
  /// grapheme and hide both the terminator and the carriage return.
  private static func splitKeepingTerminator(_ source: String)
    -> (lines: [String], terminator: Bool)
  {
    let units = Array(source.utf16)
    guard !units.isEmpty else {
      return ([], false)
    }
    let terminator = units.last == CodeUnits.lineFeed
    let body = terminator ? CodeUnits.string(units.dropLast()) : source
    return (JavaScriptString.split(body, on: CodeUnits.lineFeed), terminator)
  }

  private static func joinWithTerminator(
    _ lines: [String],
    _ terminator: Bool,
  ) -> String {
    let joined = lines.joined(separator: "\n")
    return terminator ? joined + "\n" : joined
  }
}

File: packages/core/src/markers/cli/bindingLifecycle.ts
// Shared marker-editing and registry-event plumbing for the three commands that
// write a binding: bind (place + register), unbind (remove + release), and rebind
// (move + release + re-register).
//
// It lives in one place on purpose. `use-cases` refuses a hand-placed marker and a
// hand-edited ledger, and those refusals are only fair while the tool itself is
// the single writer of BOTH halves — the marker in the source and the event in
// `bindings.jsonl`. Two implementations of "move a marker" would be two chances
// to move one half without the other.
import { BINDING_REGISTRY_SCHEMA_ID } from "../constants.js";
import { parseMarkerLine } from "../markerLine.js";
import { UCM_VERSION } from "../../version.js";
import type { MarkerFs } from "./io.js";
import { collectSourceInputs, toPosix } from "./shared.js";
import type { CommentPrefixConfig } from "../commentPrefix.js";
import { resolveCommentPrefix } from "../commentPrefix.js";

export type MarkerMode = "explicit" | "swift-func";

// Where a marker should go: a single declaration line (swift-func, marker goes
// immediately before it) or an inclusive span (explicit, a marker on each side).
export interface MarkerPlacement {
  mode: MarkerMode;
  line?: number;
  startLine?: number;
  endLine?: number;
}

export interface MarkerEditError {
  code: string;
  message: string;
}

// One marker occurrence in one file. `end_line` is null for a lone start marker
// (a Swift inferred span), which has no closing marker to remove.
export interface MarkerLocation {
  file_path: string;
  start_line: number;
  end_line: number | null;
}

export type InsertResult = { contents: string } | { error: MarkerEditError };

// Split into logical lines, remembering whether the file ended with a newline so
// an edit does not silently add or remove the trailing terminator.
function splitKeepingTerminator(source: string): { lines: string[]; terminator: boolean } {
  if (source === "") {
    return { lines: [], terminator: false };
  }
  const terminator = source.endsWith("\n");
  const body = terminator ? source.slice(0, -1) : source;
  return { lines: body.split("\n"), terminator };
}

function joinWithTerminator(lines: string[], terminator: boolean): string {
  const joined = lines.join("\n");
  return terminator ? `${joined}\n` : joined;
}

// Insert the marker(s) for `slug` into `source` at `placement`.
export function insertMarkerLines(
  source: string,
  commentPrefix: string,
  slug: string,
  placement: MarkerPlacement
): InsertResult {
  const { lines, terminator } = splitKeepingTerminator(source);
  const marker = `${commentPrefix}: @use-case:${slug}`;

  if (placement.mode === "swift-func") {
    if (placement.line === undefined || placement.line < 1) {
      return { error: { code: "BIND_LINE_REQUIRED", message: "--line is required for swift-func bind" } };
    }
    const insertAt = placement.line - 1;
    if (insertAt > lines.length) {
      return {
        error: { code: "BIND_LINE_OUT_OF_RANGE", message: `--line ${placement.line} is past end of file` }
      };
    }
    lines.splice(insertAt, 0, marker);
    return { contents: joinWithTerminator(lines, terminator) };
  }

  if (placement.startLine === undefined || placement.endLine === undefined) {
    return {
      error: {
        code: "BIND_SPAN_REQUIRED",
        message: "--start-line and --end-line are required for explicit bind"
      }
    };
  }
  if (
    placement.startLine < 1 ||
    placement.endLine < placement.startLine ||
    placement.endLine > lines.length
  ) {
    return {
      error: {
        code: "BIND_SPAN_OUT_OF_RANGE",
        message: `explicit span ${placement.startLine}-${placement.endLine} is out of range`
      }
    };
  }
  const endMarker = `${commentPrefix}: @use-case:end ${slug}`;
  // Insert end first (higher index) so the start insertion does not shift it.
  lines.splice(placement.endLine, 0, endMarker);
  lines.splice(placement.startLine - 1, 0, marker);
  return { contents: joinWithTerminator(lines, terminator) };
}

// Find `slug`'s marker lines in one file's contents.
//
// This reads the raw lines rather than the scanner's binding records on purpose:
// a marker whose span is BROKEN (the declaration after it is not a func, say)
// produces no binding record, and that is exactly the state someone needs to
// release themselves from.
export function locateMarkerLines(
  filePath: string,
  contents: string,
  commentPrefix: string,
  slug: string
): MarkerLocation | null {
  const { lines } = splitKeepingTerminator(contents);
  let startIndex: number | null = null;
  let endIndex: number | null = null;
  for (let i = 0; i < lines.length; i += 1) {
    const parse = parseMarkerLine(lines[i], commentPrefix);
    if (parse.kind === "start" && parse.slug === slug && startIndex === null) {
      startIndex = i;
      continue;
    }
    if (parse.kind === "end" && parse.slug === slug && startIndex !== null && endIndex === null) {
      endIndex = i;
    }
  }
  if (startIndex === null) {
    return null;
  }
  return {
    file_path: filePath,
    start_line: startIndex + 1,
    end_line: endIndex === null ? null : endIndex + 1
  };
}

// Remove the marker line(s) named by `location` from `contents`.
export function removeMarkerLines(contents: string, location: MarkerLocation): string {
  const { lines, terminator } = splitKeepingTerminator(contents);
  const drop = new Set<number>([location.start_line - 1]);
  if (location.end_line !== null) {
    drop.add(location.end_line - 1);
  }
  return joinWithTerminator(
    lines.filter((_line, index) => !drop.has(index)),
    terminator
  );
}

export interface FindSlugMarkerOptions {
  fs: MarkerFs;
  productRoot: string;
  slug: string;
  commentConfig?: CommentPrefixConfig;
  skipPaths?: string[];
}

// A slug's marker, wherever it is in the product tree, with the file's contents.
// Returns null when no marker for the slug exists (already removed by hand).
export function findSlugMarker(
  options: FindSlugMarkerOptions
): { location: MarkerLocation; contents: string; commentPrefix: string } | null {
  const inputs = collectSourceInputs(options.productRoot, {
    fs: options.fs,
    config: options.commentConfig,
    skipPaths: options.skipPaths
  });
  for (const input of inputs) {
    const commentPrefix = resolveCommentPrefix(input.file_path, options.commentConfig, input.contents);
    if (commentPrefix === null || commentPrefix === undefined) {
      continue;
    }
    const location = locateMarkerLines(
      toPosix(input.file_path),
      input.contents,
      commentPrefix,
      options.slug
    );
    if (location) {
      return { location, contents: input.contents, commentPrefix };
    }
  }
  return null;
}

export interface RegistryEventInput {
  command: string;
  rowId: string;
  bindingSlug: string;
  reason: string;
  eventId: string;
  createdAt: string;
  version?: string;
}

function registryEvent(
  eventType: "binding_registered" | "binding_released",
  input: RegistryEventInput
): Record<string, unknown> {
  return {
    schema: BINDING_REGISTRY_SCHEMA_ID,
    event_type: eventType,
    event_id: input.eventId,
    created_at: input.createdAt,
    created_by: {
      tool: "use-cases",
      command: input.command,
      version: input.version ?? UCM_VERSION
    },
    row_id: input.rowId,
    binding_slug: input.bindingSlug,
    reason: input.reason
  };
}

export function bindingRegisteredEvent(input: RegistryEventInput): Record<string, unknown> {
  return registryEvent("binding_registered", input);
}

// The event that ENDS a registration. Appended, never a deletion, so the ledger
// stays append-only while the binding becomes re-pointable.
export function bindingReleasedEvent(input: RegistryEventInput): Record<string, unknown> {
  return registryEvent("binding_released", input);
}
